test/bio_dgram_test.c

```diff
@@ -0,0 +1,308 @@
+     * when not enabled
+     */
+    tx_msg[0].peer  = addr2;
+    tx_msg[0].local = addr1;
+    ret = BIO_sendmmsg(b1, tx_msg, sizeof(BIO_MSG), 1, 0);
+    if (!TEST_int_eq((int)ret, -3))
+        goto err;
+
+    /* Enable local if we are using it */
+    if (BIO_dgram_get_local_addr_cap(b1) > 0 && use_local) {
+        if (!TEST_int_eq(BIO_dgram_set_local_addr_enable(b1, 1), 1))
+            goto err;
+    } else {
+        tx_msg[0].local = NULL;
+    }
+
+    /* Third effort should succeed */
+    ret = BIO_sendmmsg(b1, tx_msg, sizeof(BIO_MSG), 1, 0);
+    if (!TEST_int_eq((int)ret, 1))
+        goto err;
+
+    /* Now try receiving */
+    rx_msg[0].data      = rx_buf;
+    rx_msg[0].data_len  = sizeof(rx_buf);
+    rx_msg[0].peer      = addr3;
+    rx_msg[0].local     = addr4;
+    rx_msg[0].flags     = (1UL<<31); /* undefined flag, should be erased */
+
+    memset(rx_buf, 0, sizeof(rx_buf));
+
+    /*
+     * Should fail at first due to local being requested when not
+     * enabled
+     */
+    ret = BIO_recvmmsg(b2, rx_msg, sizeof(BIO_MSG), 1, 0);
+    if (!TEST_int_eq((int)ret, -3))
+        goto err;
+
+    /* Fields have not been modified */
+    if (!TEST_int_eq((int)rx_msg[0].data_len, sizeof(rx_buf)))
+        goto err;
+
+    if (!TEST_ulong_eq((unsigned long)rx_msg[0].flags, 1UL<<31))
+        goto err;
+
+    /* Enable local if we are using it */
+    if (BIO_dgram_get_local_addr_cap(b2) > 0 && use_local) {
+        if (!TEST_int_eq(BIO_dgram_set_local_addr_enable(b2, 1), 1))
+            goto err;
+    } else {
+        rx_msg[0].local = NULL;
+    }
+
+    /* Do the receive. */
+    ret = BIO_recvmmsg(b2, rx_msg, sizeof(BIO_MSG), 1, 0);
+    if (!TEST_int_eq((int)ret, 1))
+        goto err;
+
+    /* data_len should have been updated correctly */
+    if (!TEST_int_eq((int)rx_msg[0].data_len, 5))
+        goto err;
+
+    /* flags should have been zeroed */
+    if (!TEST_int_eq((int)rx_msg[0].flags, 0))
+        goto err;
+
+    /* peer address should match expected */
+    if (!TEST_int_eq(compare_addr(addr3, addr1), 1))
+        goto err;
+
+    if (rx_msg[0].local != NULL) {
+        /* If we are using local, it should match expected */
+        if (!TEST_int_eq(compare_addr(addr4, addr2), 1))
+            goto err;
+    }
+
+    testresult = 1;
+err:
+    BIO_free(b1);
+    BIO_free(b2);
+    if (fd1 >= 0)
+        BIO_closesocket(fd1);
+    if (fd2 >= 0)
+        BIO_closesocket(fd2);
+    BIO_ADDR_free(addr1);
+    BIO_ADDR_free(addr2);
+    BIO_ADDR_free(addr3);
+    BIO_ADDR_free(addr4);
+    return testresult;
+}
+
+static int test_bio_dgram(void)
+{
+    /* Test without local */
+    if (test_bio_dgram_impl(AF_INET, 0) < 1)
+        return 0;
+
+    if (test_bio_dgram_impl(AF_INET6, 0) < 1)
+        return 0;
+
+    /* Test with local */
+    if (test_bio_dgram_impl(AF_INET, 1) < 1)
+        return 0;
+
+    if (test_bio_dgram_impl(AF_INET6, 1) < 1)
+        return 0;
```

Comment:
Usually this would be done as multiple testcases test - i.e. switch() by `int idx` param.